pkg/pool-linear/contracts/beefy/BeefyLinearPool.sol

```diff
@@ -0,0 +1,91 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/pool-linear/IBeefyVault.sol";
+
+import "../LinearPool.sol";
+
+contract BeefyLinearPool is LinearPool {
+    IBeefyVault private immutable _tokenVault;
+
+    uint256 private immutable _balanceScaleFactor;
+
+    struct ConstructorArgs {
+        IVault vault;
+        string name;
+        string symbol;
+        IERC20 mainToken;
+        IERC20 wrappedToken;
+        address assetManager;
+        uint256 upperTarget;
+        uint256 swapFeePercentage;
+        uint256 pauseWindowDuration;
+        uint256 bufferPeriodDuration;
+        address owner;
+    }
+
+    constructor(ConstructorArgs memory args)
+        LinearPool(
+            args.vault,
+            args.name,
+            args.symbol,
+            args.mainToken,
+            args.wrappedToken,
+            args.upperTarget,
+            _toAssetManagerArray(args),
+            args.swapFeePercentage,
+            args.pauseWindowDuration,
+            args.bufferPeriodDuration,
+            args.owner
+        )
+    {
+        IBeefyVault tokenVault = IBeefyVault(address(args.wrappedToken));
+
+        _tokenVault = tokenVault;
+        address want = tokenVault.want();
+
+        // Beefy vault tokens are always 18 decimal places, but are stored in the precision of the mainToken
+        // ie: If ppfs === 1e18, 1 USDC === 0.000_000_000_001_000_000 mooUSDC
+        // ie: If ppfs === 1e18, 1 DAI === 1 mooDAI
+        // ie: If ppfs === 1e18, 1 WBTC === 0.000_000_000_100_000_000 mooWBTC
+        // -----------
+        // Internally, the LinearPool scales all balances and rates up to 18 decimal places, meaning that 1 USDC is
+        // represented as 1e18 by the LinearPool. Since the mooUSDC is already 18 decimals,
+        // but in a different representation, we need to account for that in our wrappedTokenRate.
+        // Since we only accept tokens with <= 18 decimals, we know the smallest this can be is 10^0 === 1
+        _balanceScaleFactor = 10**(SafeMath.add(18, SafeMath.sub(18, ERC20(want).decimals())));
+
+        _require(address(args.mainToken) == want, Errors.TOKENS_MISMATCH);
+    }
+
+    function _toAssetManagerArray(ConstructorArgs memory args) private pure returns (address[] memory) {
+        // We assign the same asset manager to both the main and wrapped tokens.
+        address[] memory assetManagers = new address[](2);
+        assetManagers[0] = args.assetManager;
+        assetManagers[1] = args.assetManager;
+
+        return assetManagers;
+    }
+
+    function _getWrappedTokenRate() internal view override returns (uint256) {
+        uint256 vaultTotalSupply = _tokenVault.totalSupply();
+        return
+            vaultTotalSupply == 0
+                ? _balanceScaleFactor
+                : (_tokenVault.balance() * _balanceScaleFactor) / vaultTotalSupply;
```

Comment:
This seems a bit odd, under which conditions will the supply be zero?

I can't quite comment on the correctness of the rest of this function, as that requires reading the Beefy code thoroughly. In particular, I'd be concerned with the possibility of reentrancy attacks manipulating the values of `totalSupply()` and `balance()` so that they're potentially out of sync.

Reply:
Ya, it should never be 0 unless everyone has left the vault. 

Not sure where the reentrancy would occur? We have protection against it on our deposits. PPFS is calculated in the same way though it ends up rounding. So using this method would have better precision.

Reply:
Not sure where reentrancy might happen as that'd depend on the details of the `tokenVault`, but any action that results in callbacks after either supply or balance is updated but before the other is also updated would result in issues. If you're convinced that can't happen, then things should be fine.

Reply:
Here is a link to the V6 code which would be the code for the live vaults. 

https://github.com/beefyfinance/beefy-contracts/blob/master/contracts/BIFI/vaults/BeefyVaultV6.sol

Balance refers to both the strategy balance and anything that remains in the vault unused. The deposit function uses a nonreentrant modifier, the only callback would have to be from the transfer of the want. The shares are minted at the end of the function, after the transfer which is what increases the total supply. 

The withdraw function burns and reduces shares first before transferring any want back to the user.